router/dataplane.go

```diff
@@ -442,105 +457,303 @@ func (d *DataPlane) Run(ctx context.Context) error {
 
-				_, err = result.OutConn.WriteBatch(writeMsgs, syscall.MSG_DONTWAIT)
-				if err != nil {
-					var errno syscall.Errno
-					if !errors.As(err, &errno) ||
-						!(errno == syscall.EAGAIN || errno == syscall.EWOULDBLOCK) {
-						log.Debug("Error writing packet", "err", err)
-						// error metric
-					}
-					inputCounters.DroppedPacketsTotal.Inc()
-					continue
-				}
-				// ok metric
-				outputCounters := d.forwardingMetrics[result.EgressID]
-				outputCounters.OutputPacketsTotal.Inc()
-				outputCounters.OutputBytesTotal.Add(float64(len(result.OutPkt)))
+	hasher := fnv.New32a()
+	hasher.Write(d.randomValue)
+	hasher.Write(tmpBuffer[:])
+	hasher.Write(data[slayers.CmnHdrLen : slayers.CmnHdrLen+addrHdrLen+1])
+	return hasher.Sum32() % d.procRoutines, nil
+}
+
+func (d *DataPlane) returnPacket(pkt *packet) {
+	d.packetPool <- pkt
+}
+
+func (d *DataPlane) initProcessingRoutine(id int) error {
+	log.Debug("Initialize processing routine with", "id", id)
+	c := d.procChannels[id]
+	processor := newPacketProcessor(d, 0)
+	var scmpErr scmpError
+	for d.running {
+		p := <-c
+		processor.ingressID = p.ingress
+		metrics := d.forwardingMetrics[p.ingress]
+		result, err := processor.processPkt(p.rawPacket, p.srcAddr)
+		egress := result.EgressID
+		switch {
+		case err == nil:
+		case errors.As(err, &scmpErr):
+			if !scmpErr.TypeCode.InfoMsg() {
+				log.Debug("SCMP", "err", scmpErr)
 			}
+			// SCMP go back the way they came.
+			egress = p.ingress
+		default:
+			log.Debug("Error processing packet", "err", err)
+			metrics.DroppedPacketsTotal.Inc()
+			d.returnPacket(p)
+			continue
+		}
+		if result.OutConn == nil { // e.g. BFD case no message is forwarded
+			d.returnPacket(p)
+			continue
+		}
+		fwCh, ok := d.forwardChannels[egress]
+		if !ok {
+			log.Debug("Error determining forwarder. Egress is invalid", "egress", egress)
+			metrics.DroppedPacketsTotal.Inc()
+			d.returnPacket(p)
+			continue
 		}
+		p.rawPacket = result.OutPkt
+		if result.OutAddr != nil {
+			p.dstAddr = result.OutAddr
+		}
+
+		select {
+		case fwCh <- p:
+			// TODO(rohrerj) add metrics
+		default:
+			d.returnPacket(p)
+			metrics.DroppedPacketsTotal.Inc()
+		}
+
 	}
+	return nil
+}
 
-	for k, v := range d.bfdSessions {
-		go func(ifID uint16, c bfdSession) {
-			defer log.HandlePanic()
-			if err := c.Run(ctx); err != nil && err != bfd.AlreadyRunning {
-				log.Error("BFD session failed to start", "ifID", ifID, "err", err)
+func (d *DataPlane) initForwarder(ni NetworkInterface) error {
+	log.Debug("Initialize forwarder for", "interface", ni.InterfaceId)
+	c := d.forwardChannels[ni.InterfaceId]
+	writeMsgs := make(underlayconn.Messages, d.interfaceBatchSize)
+	for i := range writeMsgs {
+		writeMsgs[i].Buffers = make([][]byte, 1)
+	}
+	metrics := d.forwardingMetrics[ni.InterfaceId]
+	currentPackets := make([]*packet, d.interfaceBatchSize)
+	byteLen := 0
+	for d.running {
+		p := <-c
+		writeMsgs[0].Buffers[0] = p.rawPacket
```

Comment:
I think this could be simplified as follows (not tested though):

```
for d.running {
    select {
    case p := <-c:
        writeMsgs[0].Buffers[0] = p.rawPacket
        ...
        for i := 0; i < min(len(c), d.interfaceBatchSize - 1); i++ {
           p := <-c
           writeMsgs[i].Buffers[0] = p.rawPacket
           ...
        }
        ni.Conn.WriteBatch(writeMsgs[:i], syscall.MSG_DONTWAIT)
        ...
    }
}
```

Reply:
Also, can you create a function for the following:

```
writeMsgs[i].Buffers[0] = p.rawPacket
writeMsgs[i].Addr = nil
if p.dstAddr != nil {
	writeMsgs[i].Addr = p.dstAddr
}
```

Reply:
> I think this could be simplified as follows (not tested though):
> 
> ```
> for d.running {
>     select {
>     case p := <-c:
>         writeMsgs[0].Buffers[0] = p.rawPacket
>         ...
>         for i := 0; i < min(len(c), d.interfaceBatchSize - 1); i++ {
>            p := <-c
>            writeMsgs[i].Buffers[0] = p.rawPacket
>            ...
>         }
>         ni.Conn.WriteBatch(writeMsgs[:i], syscall.MSG_DONTWAIT)
>         ...
>     }
> }
> ```

This code will not work. For example we have len(c) = 2 directly before the loop. Now you read a message from the channel, which decreases len(c) and increment i. Now you have the condition 1 < min(1, 256), which is false although the channel is not empty yet.

We would have to check whether len(c) != 0. This would lead to a race condition with the writers but since we have only a single reader for that channel and it's not a problem whether we write that message in the current or next batch, that would probably work.

Reply:
Done